overlord/devicestate/devicemgr.go

```diff
@@ -0,0 +1,384 @@
+	runner.AddHandler("deliver-serial-request", doDeliverSerialRequest, nil)
+	runner.AddHandler("retrieve-n-install-serial", doRetrieveAndInstallSerial, nil)
+
+	return &DeviceManager{state: s, runner: runner}, nil
+}
+
+func (m *DeviceManager) ensureOperational() error {
+	m.state.Lock()
+	defer m.state.Unlock()
+	// XXX: auth.Device/SetDevice should probably move to devicestate
+	// they are not quite just about auth and also we risk circular imports
+	// (auth will need to mediate bits from devicestate soon)
+	devIdentity, err := auth.Device(m.state)
+	if err != nil {
+		return err
+	}
+
+	if devIdentity.Serial != "" {
+		// serial is set, we are all set
+		return nil
+	}
+
+	if devIdentity.Brand == "" || devIdentity.Model == "" {
+		// need first-boot, loading of model assertion info
+		if release.OnClassic {
+			// XXX: cheat for now to get us started somewhere
+			devIdentity.Brand = "canonical"
+			devIdentity.Model = "pc"
+			err := auth.SetDevice(m.state, devIdentity)
+			if err != nil {
+				return err
+			}
+			m.state.Unlock()
+			m.state.Lock()
+		} else {
+			// full first-boot stuff!
+			// TODO: move first boot setup/invocation here!
+			panic("need full first-boot to initialize brand and model of device")
+		}
+	}
+
+	for _, chg := range m.state.Changes() {
+		if chg.Kind() == "become-operational" && !chg.Status().Ready() {
+			// change already in motion
+			return nil
+		}
+	}
+
+	// XXX: some of these will need to be split and use hooks
+	// retries might need to embrace more than one "task" then,
+	// need to be careful
+
+	genKey := m.state.NewTask("generate-device-key", i18n.G("Generate device key"))
+	serialReq := m.state.NewTask("deliver-serial-request", i18n.G("Deliver device serial request"))
+	serialReq.WaitFor(genKey)
+	retrieveSerial := m.state.NewTask("retrieve-n-install-serial", i18n.G("Retrieve and install device serial"))
+	retrieveSerial.WaitFor(serialReq)
+	serialReq.Set("serial-setup-task", retrieveSerial.ID())
+
+	chg := m.state.NewChange("become-operational", i18n.G("Setting up device identity"))
+	chg.AddAll(state.NewTaskSet(genKey, serialReq, retrieveSerial))
+
+	return nil
+}
+
+// Ensure implements StateManager.Ensure.
+func (m *DeviceManager) Ensure() error {
+	err := m.ensureOperational()
+	if err != nil {
+		return err
+	}
+	m.runner.Ensure()
+	return nil
+}
+
+// Wait implements StateManager.Wait.
+func (m *DeviceManager) Wait() {
+	m.runner.Wait()
+}
+
+// Stop implements StateManager.Stop.
+func (m *DeviceManager) Stop() {
+	m.runner.Stop()
+}
+
+var (
+	keyLength = 4096
+	// XXX: a 2nd different URL for nonce?
+	// TODO: this will come as config from the gadget snap
+	serialRequestURL = "https://serial.request" // XXX dummy value!
+)
+
+func doGenerateDeviceKey(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+	// XXX: where to store device private key ultimately? use asserts support?
+	var encoded string
+	err := st.Get("device-keypair", &encoded)
```

Comment:
Let's please not use the state, though. This is bulky data that pretty much never changes, so a bad fit.